### DISCUSSION TITLE
Get raw request body in pb_hook?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#4098] [Q&A] [answered]
I'm trying to use Amazon SNS but am having trouble getting the POST body. It's supposed to be a JSON payload, but `JSON.stringify(requestInfo(c).data)` just shows `{}`. I did some postman tests and discovered that the `content-type` header needs to be `application/json` before it will work. Amazon sends `text/plain`, so is there a way to just get the raw POST body?

According to https://docs.aws.amazon.com/sns/latest/dg/SendMessageToHttp.prepare.html

```
POST / HTTP/1.1
    x-amz-sns-message-type: SubscriptionConfirmation
    x-amz-sns-message-id: 165545c9-2a5c-472c-8df2-7ff2be2b3b1b
    x-amz-sns-topic-arn: arn:aws:sns:us-west-2:123456789012:MyTopic
    Content-Length: 1336
    Content-Type: text/plain; charset=UTF-8
    Host: example.com
    Connection: Keep-Alive
    User-Agent: Amazon Simple Notification Service Agent
    
{
  "Type" : "SubscriptionConfirmation",
  "MessageId" : "165545c9-2a5c-472c-8df2-7ff2be2b3b1b",
  "Token" : "2336412f37f...",
  "TopicArn" : "arn:aws:sns:us-west-2:123456789012:MyTopic",
  "Message" : "You have chosen to subscribe to the topic arn:aws:sns:us-west-2:123456789012:MyTopic.\nTo confirm the subscription, visit the SubscribeURL included in this message.",
  "SubscribeURL" : "https://sns.us-west-2.amazonaws.com/?Action=ConfirmSubscription&TopicArn=arn:aws:sns:us-west-2:123456789012:MyTopic&Token=2336412f37...",
  "Timestamp" : "2012-04-26T20:45:04.751Z",
  "SignatureVersion" : "1",
  "Signature" : "EXAMPLEpH+...",
  "SigningCertURL" : "https://sns.us-west-2.amazonaws.com/SimpleNotificationService-f3ecfb7224c7233fe7bb5f59f96de52f.pem"
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand what is the issue but you can check https://pocketbase.io/docs/js-routing/#reading-request-body.

Alternatively, if you really want to read the raw body you can try with:

```js
// note: it can be read only once
const raw = readerToString(c.request().body)
```

**Reply by benallfree** [reply]
Just closing the loop on this: the bottom line is that a POST request with a content-type of `text/plain` will NOT parse the body as JSON, so `$apis.requestInfo(c).data` will be empty (`{}`).

I ran into this because for whatever reason, Amazon SNS sends a JSON body but the content type is `text/plain`. Using `const data = JSON.parse(c.request().body)` worked.
